### ISSUE TITLE
Azure Databases for MySQL, cipher suites not supported

### ISSUE THREAD

**Issue by SumBeam** [Sequel-Ace/Sequel-Ace#419]
- Sequel Ace Version: 2.2.1
- macOS Version: 10.15.4
- MySQL Version: 8.0

**Description**
We are going to use Azure Databases and we noticed that we can't connect with Sequel Ace to these machines.

We are getting this error while connecting: ![Screenshot 2020-10-21 at 08 51 41](https://user-images.githubusercontent.com/1443527/96683762-e0991a00-137a-11eb-884e-e7ba9253d1a5.png)

We have configured them with a minimal TLS version of 1.2. Based on the documention they accept these 4 cipher suites:

- TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384
- TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256
- TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA384
- TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256

![image](https://user-images.githubusercontent.com/1443527/96685017-d24bfd80-137c-11eb-95e6-033541c18c24.png)

It looks like those cipher suites are not supported by Sequel Ace? I couldn't find them in the preferences tab.
More information at Microsoft: https://docs.microsoft.com/en-us/azure/mysql/concepts-ssl-connection-security

**Steps To Reproduce**
1. Create a database @ Microsoft
2. Try to connect

**Expected Behaviour**
That we are able to connect to our database.

**Is Issue Present in [Latest Beta](https://github.com/Sequel-Ace/Sequel-Ace/releases)?**
Looks like there is no beta that is newer then the 2.2.1 release from the App Store

Please let me know if you need more information

**Comment by Jason-Morcos** [comment]
Possibly tangentially related to #164

**Comment by SumBeam** [comment]
Hi,

Thanks for looking into this issue. I saw that the PR #428 was merged to the main branch. I'm happy to test this feature to see if that fixes this issue.

Let me know if you need any help 👍

**Comment by Jason-Morcos** [comment]
> Hi,
> 
> Thanks for looking into this issue. I saw that the PR #428 was merged to the main branch. I'm happy to test this feature to see if that fixes this issue.
> 
> Let me know if you need any help 👍

Awesome! Yes! Could you check out the latest beta and let us know if the issue is fixed?
https://github.com/Sequel-Ace/Sequel-Ace/releases/tag/2.3.0-beta1

**Comment by SumBeam** [comment]
Hey @Jason-Morcos 

Sure. I've downloaded the latest beta but I still see the same error from the server.

I've confirmed that the new suites are added to the list in the preferences (3rd and 4th option + the other 2 are in the list but outside the screenshot area).

![image](https://user-images.githubusercontent.com/1443527/97844475-61e19c80-1ceb-11eb-9f96-9030b3cd3918.png)

Is there maybe any additional logging I can enable?

**Comment by Jason-Morcos** [comment]
Looking at your original issue screenshot, I see SSL turned on but no key file or cert selected. As I understand it, don't you have to have a key file to connect via SSL?

I think this is now a connection settings issue and not a cipher issue. Perhaps someone in @Sequel-Ace/all has some insight! :)

**Comment by SumBeam** [comment]
No, we do not need a SSL key file to connect :) 

I tried something and I moved up the 4th SSL connection (the one MySQLWorkbench used) to the top and now I was able to connect. So maybe it has something to do with the order of the items in the preferences?

**Comment by Jason-Morcos** [comment]
> No, we do not need a SSL key file to connect :) 
> 
> 
> 
> I tried something and I moved up the 4th SSL cipher suite (the one MySQLWorkbench used) to the top and now I was able to connect. So maybe it has something to do with the order of the items in the preferences?
> 
> 
> 
> <img width="593" alt="Screenshot 2020-11-02 at 09 30 41" src="https://user-images.githubusercontent.com/1443527/97846312-1b417180-1cee-11eb-8859-1a4296600b7c.png">
> 
> 

Interesting! Was that screenshot you sent the order that actually worked? Which cipher did you move to the top?

**Comment by SumBeam** [comment]
@Jason-Morcos I thought I figured it out with moving it to the top, but I wanted to confirm your statement. I restored the settings to default and even then I was able to connect..

And now when I disable the checkmark 'Connect using SSL' and/or move all the cipher suites below the 'Disabled cipher suites' mark, it is connecting fine... 🤐 

So I don't know whats happening right now. It clearly wasn't working when I launched the app yesterday morning for the first time and after I tried some stuff with the order it started to work magically. And now I can't get it back into a broken state.

![image](https://user-images.githubusercontent.com/1443527/97967582-b35a5c00-1dbd-11eb-921e-97f59c59ec0f.png)

**Comment by SumBeam** [comment]
**TLDR: To come back to your original question. Yes, I need to move up 1 of the accepted cipher suites to the top of the list to be able to connect.**


Strange, after a reboot of my system and waited for about 30 minutes I tried it again... With an empty Cipher Suites list (everything moved below the disabled marker) it throws an error '_MySQL said: SSL connection error: Failed to set ciphers to use_'

When I restore that list to default, it throws this error: '_MySQL said: FIPS cipher suites are enforced for this server. Please specify FIPS complying cipher suites in your SSL/TLS settings._'

When I move up the ECDHE-RSA-AES128-GCM-SHA256 (or 1 of the other accepted ciphers) to the top, it connects just fine. After a successful connection I am unable to get it into a broken state again by redoing the above steps. Looks like openssl is caching the connections or something?

**Comment by Jason-Morcos** [comment]
Thank you for the info!! Will likely tweak to move `ECDHE-RSA-AES128-GCM-SHA384` to the top and `ECDHE-RSA-AES128-GCM-SHA256` to second in the cipher list

**Comment by Jason-Morcos** [comment]
Actually, it looks like our top cipher `ECDHE-RSA-AES256-GCM-SHA384` should be FIPS compliant. I think for now we can assume this is fine as is and tweak only if necessary in the future? In any case, I'm glad we were able to get it working for you!! Caching is wild lol

**Comment by SumBeam** [comment]
Cool, thanks for the change! When the release is out I can reach out to my team to start using this product and moving away from Sequel Pro :)

But isn't it weird behaviour that I need to move up/down the ciphers? As a user I would expect that it is trying from top to bottom which one is working.. (but as a developer I know that systems are not working like that haha)

**Comment by Jason-Morcos** [comment]
> Cool, thanks for the change! When the release is out I can reach out to my team to start using this product and moving away from Sequel Pro :)
> 
> 
> 
> But isn't it weird behaviour that I need to move up/down the ciphers? As a user I would expect that it is trying from top to bottom which one is working.. (but as a developer I know that systems are not working like that haha)

It certainly is weird! I'm hoping it was some strange caching issue specific to your local system, but definitely let us know if the rest of your team ends up having to do the cipher order shuffle! Because it's definitely not optimal behavior haha

**Comment by superpowered** [comment]
@Jason-Morcos Wanted to pop in here to note that this bug still exists, but this workaround also still works. 
Just quick shuffling the top two ciphers allows it to work. (even if they get shuffled back the other way after the connection) 
My guess is once the connection works its caching or saving the cipher type somewhere allowing to to work from then on out. But there's still the bug of it having an issue if the `ECDHE-RSA-AES256-GCM-SHA384` is first when `ECDHE-RSA-AES128-GCM-SHA256` is the type needed on initial connection.

This was when connecting to an Azure Flexible MySQL Server, with a CA cert required.
